### PR TITLE
feat(ops): V8 Fast Calls

### PR DESCRIPTION
**Pull request by littledivy**
Relands auto fast api template generation for ops. This optimization is opt in:

```rust
#[op(fast)]
fn something_mhmm(
  state: &mut OpState
) { 
  // mhmm
}
```

```js
const { something_mhmm } = core.ops;
something_mhmm.fast(); // something_mhmm
```

### CODE REVIEW

**Comment by aapoalas** [comment]
Any idea if this can be merged now or is it still flaking?

**Comment by aapoalas** [comment]
Regarding the `op_func.call(...args)` usage: Would the normal calling convention of `ops.op_func(...args)` work if the `ops` object was created as an `External` from the `OpCtx`? (IIRC it's that pointer that we need for ops, right?) ie. Does the "this" object matter for Fast API data receiving?

I would presume it does not matter and I remember reading something about how there's a "Holder" object for Fast API functions which is not the same thing as the "this" argument which is sometimes called "holder" in V8 code. Anyway, there are two possibilities that I can see:
1. Either the `this` argument is relevant, in which case `ops` should be created as a `v8::External` to retain the `ops.op_func(...args)` calling convention.
2. Or (more likely) the `this` argument is irrelevant, in which case I would presume that Fast Calls get the `data` object anyway, and the `ops.op_func.call(...args)` calling convention is unnecessary.

@divy I presume you've tested that the `.call()` method usage is truly necessary? How hard would it be to test the `ops` as `v8::External` case?

File: cli/bench/deno_common.js [hunk 1]
```diff
@@ -5,12 +5,31 @@ Deno.bench("date_now", { n: 5e5 }, () => {
   Date.now();
 });
 
+// Fast API calls
+{
+  // deno-lint-ignore camelcase
+  const { op_add } = Deno.core.ops;
+  // deno-lint-ignore no-inner-declarations
+  function add(a, b) {
+    return op_add.call(a, b);
```
Comment (bartlomieju) [inline cli/bench/deno_common.js:14]: change to `op_add.fast()`

File: cli/bench/deno_common.js [hunk 2]
```diff
@@ -5,12 +5,31 @@ Deno.bench("date_now", { n: 5e5 }, () => {
...
+  }
+  // deno-lint-ignore no-inner-declarations
+  function addJS(a, b) {
+    return a + b;
+  }
+  Deno.bench("op_add", { n: 1e7 }, () => add(1, 2));
```
Comment (bartlomieju) [inline cli/bench/deno_common.js:20]: Remove these second argument objects, `n` is ignored anyway

File: core/bindings.rs [hunk 3]
```diff
@@ -9,17 +9,38 @@ use crate::modules::ModuleMap;
...
+      if let Some(fast_fn) = &ctx.decl.fast_fn {
+        references.push(v8::ExternalReference {
+          pointer: fast_fn.function() as _,
+        });
+      }
+    }
+  }
+
+  let refs = v8::ExternalReferences::new(&references);
+  std::mem::forget(references);
```
Comment (bartlomieju) [inline core/bindings.rs:41]: Add a comment here that it doesn't leak, because `v8` takes ownership

File: core/bindings.rs [hunk 4]
```diff
@@ -94,18 +116,53 @@ pub fn initialize_context<'s>(
...
+
+      let method_obj = object_template.new_instance(scope).unwrap();
+      method_obj.set_aligned_pointer_in_internal_field(
+        crate::runtime::V8_WRAPPER_OBJECT_INDEX,
+        ctx_ptr,
+      );
+      set_func_raw(
+        scope,
+        method_obj,
+        "call",
```
Comment (bartlomieju) [inline core/bindings.rs:147]:
```suggestion
        "fast",
```

File: core/bindings.rs [hunk 5]
```diff
@@ -94,18 +116,53 @@ pub fn initialize_context<'s>(
...
   scope: &mut v8::HandleScope,
   ops_obj: v8::Local<v8::Object>,
   op_ctxs: &[OpCtx],
+  snapshot_loaded: bool,
 ) {
   for ctx in op_ctxs {
     let ctx_ptr = ctx as *const OpCtx as *const c_void;
-    set_func_raw(scope, ops_obj, ctx.decl.name, ctx.decl.v8_fn_ptr, ctx_ptr);
+
+    if ctx.decl.fast_fn.is_some() && snapshot_loaded {
```
Comment (bartlomieju) [inline core/bindings.rs:133]: This conditional could use a comment what's going on here

File: core/bindings.rs [hunk 6]
```diff
@@ -129,13 +186,24 @@ pub fn set_func_raw(
...
-    .unwrap();
+  let builder =
+    v8::FunctionTemplate::builder_raw(callback).data(external.into());
+  let templ = if let Some(fast_function) = fast_function {
+    if !snapshot_loaded {
+      builder.build(scope)
+    } else {
+      // TODO(@littledivy): Support fast api overloads in ops.
+      builder.build_fast(scope, &**fast_function, None)
+    }
```
Comment (bartlomieju) [inline core/bindings.rs:202]: Add a comment here, why we have different method used depending if snapshots was loaded

File: core/extensions.rs [hunk 7]
```diff
@@ -2,21 +2,24 @@
...
+use v8::fast_api::FastFunction;
 
 pub type SourcePair = (&'static str, &'static str);
 pub type OpFnRef = v8::FunctionCallback;
 pub type OpMiddlewareFn = dyn Fn(OpDecl) -> OpDecl;
 pub type OpStateFn = dyn Fn(&mut OpState) -> Result<(), Error>;
 pub type OpEventLoopFn = dyn Fn(Rc<RefCell<OpState>>, &mut Context) -> bool;
 
-#[derive(Clone, Copy)]
+pub trait FastFunctionSignature {}
```
Comment (bartlomieju) [inline core/extensions.rs:13]: Why is it needed? Can you add a comment?

File: core/extensions.rs [hunk 8]
```diff
@@ -32,7 +35,7 @@ impl OpDecl {
 #[derive(Default)]
 pub struct Extension {
   js_files: Option<Vec<SourcePair>>,
-  ops: Option<Vec<OpDecl>>,
+  pub ops: Option<Vec<OpDecl>>,
```
Comment (bartlomieju) [inline core/extensions.rs:38]: Why make it public?

File: core/runtime.rs [hunk 9]
```diff
@@ -233,6 +233,7 @@ fn v8_init(
     " --harmony-import-assertions",
     " --no-validate-asm",
     " --turbo_fast_api_calls",
+    " --allow-natives-syntax",
```
Comment (bartlomieju) [inline core/runtime.rs:236]: Seems like a mistake? We don't want to expose native syntax to all users

File: core/runtime.rs [hunk 10]
```diff
@@ -341,12 +341,13 @@ impl JsRuntime {
       .collect::<Vec<_>>()
       .into_boxed_slice();
 
+    let refs = bindings::external_references(&op_ctxs, !options.will_snapshot);
+    let refs: &'static v8::ExternalReferences = Box::leak(Box::new(refs));
```
Comment (bartlomieju) [inline core/runtime.rs:345]: Add a comment why we are leaking

File: core/extensions.rs (hunk 7 above)
Comment (littledivy) [reply core/extensions.rs]: removed

**Comment by aapoalas** [comment]
I finally managed to understand that the fast bindings are generated to an extra function bound into the original function but... Why? Does the references snapshotting somehow not work with direct binding to the normal op call function?

File: core/bindings.rs [hunk 11]
```diff
@@ -9,17 +9,39 @@ use crate::modules::ModuleMap;
...
+  snapshot_loaded: bool,
+) -> v8::ExternalReferences {
+  let mut references = vec![v8::ExternalReference {
+    function: call_console.map_fn_to(),
+  }];
+
+  for ctx in ops {
+    let ctx_ptr = ctx as *const OpCtx as _;
+    references.push(v8::ExternalReference { pointer: ctx_ptr });
+    references.push(v8::ExternalReference {
```
Comment (aapoalas) [inline core/bindings.rs:28]: Why are all this external refs now suddenly needed when previously they were not?

File: core/bindings.rs [hunk 12]
```diff
@@ -94,18 +117,55 @@ pub fn initialize_context<'s>(
...
   scope: &mut v8::HandleScope,
   ops_obj: v8::Local<v8::Object>,
   op_ctxs: &[OpCtx],
+  snapshot_loaded: bool,
 ) {
   for ctx in op_ctxs {
     let ctx_ptr = ctx as *const OpCtx as *const c_void;
-    set_func_raw(scope, ops_obj, ctx.decl.name, ctx.decl.v8_fn_ptr, ctx_ptr);
+
+    // If this is a fast op, we don't want it to be in the snapshot.
```
Comment (aapoalas) [inline core/bindings.rs:134]: While this explains why the boolean is used here, it does not actually explain why the fast op is not wanted in the snapshot. At least I would like to understand why the fast op is left out: Is it because of the doubling of the `OpCtx` and slow call function pointer referencing?

File: ext/flash/lib.rs [hunk 13]
```diff
@@ -1344,17 +1344,17 @@ async fn op_flash_next_async(
 // the ContextScope creation is optimized away and the op is as simple as:
 //   f(info: *const v8::FunctionCallbackInfo) { let rv = ...; rv.set_uint32(op_flash_next()); }
 #[op]
-fn op_flash_next(op_state: &mut OpState) -> u32 {
-  let flash_ctx = op_state.borrow_mut::<FlashContext>();
+fn op_flash_next(state: &mut OpState) -> u32 {
```
Comment (aapoalas) [inline ext/flash/lib.rs:1347]: Unrelated changes?

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by ry** [comment]
This is awesome - great work @littledivy
